File: medium/515/515.py
# msg = 'hello world'
# print(msg)
# Definition for a binary tree node.
# from typing import Optional
from ast import List
import sys
from typing import Deque, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
      # if not root:
      #   return root
      queue,res =Deque([root]), []
      if root:
        queue.append(root)
      
      while queue:
        max = -sys.maxsize-1
        for i in range(len(queue)):
          shift = queue.popleft()
          max = max(max,shift.val)
          if shift.left:
            queue.append(shift.left)
          if shift.right:
            queue.append(shift.right)
        res.append(max)
      return res

File: medium/515/515.ts
// Definition for a binary tree node.
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}
function largestValues(root: TreeNode | null): number[] {
  const res: number[] = []
  const queue: TreeNode[] = []
  if (root)
    queue.push(root)
  while (queue.length !== 0) {
    const size = queue.length
    let max = queue[0].val
    for (let i = 0; i < size; i++) {
      const shift = queue.shift()
      max = Math.max(max, shift?.val as number)
      if (shift?.left)
        queue.push(shift.left)

      if (shift?.right)
        queue.push(shift.right)
    }
    res.push(max)
  }
  return res
}

export {
  largestValues,
}
